### PR TITLE
feat: add scalar subquery pushdown to scan

### PR DESCRIPTION
**Pull request by parthchandra**
## Which issue does this PR close?
Part of #372  and #551 

## Rationale for this change
With Spark 4.0, the `SubquerySuite` in Spark fails as Comet scan did not support the scala subquery feature.

## What changes are included in this PR?
Adds the support for scalar subquery pushdown into Comet scan 

## How are these changes tested?
Existing Spark/sql unit tests in `SubquerySuite`

### CODE REVIEW

**Comment by parthchandra** [comment]
Currently DRAFT to make sure ci for older versions passes.

Note the shims for the older versions had to be refactored as this required a change specifically for Spark 3.3 which was different from the change required for Spark 3.4 and above.

**Comment by parthchandra** [comment]
@kazuyukitanimura Ready for your review.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala [hunk 1]
```diff
@@ -134,10 +135,42 @@ case class CometScanExec(
...
+    scalarSubqueryReplaced
+      .filterNot(_.references.exists {
+        isFileSourceConstantMetadataAttribute
+      })
+      .flatMap(DataSourceStrategy.translateFilter(_, supportNestedPredicatePushdown))
+  }
+
+  @transient
+  private lazy val pushedDownFilters =
+    translateToV1Filters(dataFilters, q => convertScalarSubqueryToLiteral(q))
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala:161]: Is it possible to define `pushedDownFilters` in Shims instead?
We can do the old way for Spark 3.x

For Spark 4.0, we can avoid the reflection like `convertScalarSubqueryToLiteral`

**Review by parthchandra — COMMENTED** [review]
@kazuyukitanimura completely refactored this so now the additional shim classes for pre-3.5 are gone and the change is now really simple.

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

**Review by andygrove — APPROVED** [review]

**Review by comphead — APPROVED** [review]
lgtm thanks @parthchandra

**Comment by parthchandra** [comment]
@kazuyukitanimura @andygrove @comphead Can we merge this?

**Comment by kazuyukitanimura** [comment]
Merged, thanks @parthchandra @andygrove @comphead

**Pull request merged**